### DISCUSSION TITLE
Injections within template

### DISCUSSION THREAD

**Discussion by aliaksei-karneyeu** [nvim-treesitter/nvim-treesitter#1917] [Q&A] [answered]
Hey guys,

I'm working a lot with helm templates that are generally golang templates with yaml syntax. I installed https://github.com/ngalaiko/tree-sitter-go-template and it hightlight properly variables, what am I missing now - yaml syntax highlight.

My question is general. How to configure more intelligent way to highlight one file type with another, like go template with yaml, erb with html and so on?

![image](https://user-images.githubusercontent.com/72257065/137121967-ecc237bf-241a-43a6-bc62-4d35370dca59.png)

**Comment by stsewd** [comment]
Take a look at https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#injections and some examples at https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/html_tags/injections.scm or similar.

Also, I'm not sure if this is yaml inside gotemplates or gotemplates inside yaml?

**Reply by aliaksei-karneyeu** [reply]
Thanks for the links.

I'm noob in treesitter and in programming, but anyway I would like to learn it little bit more by configuring highlighting for helm files.

So here are the things that I understand so far:
* `{{ .Release.Name }}` - should be highlighted by go template parser and everything else should be highlighted by yaml parser.
* yaml itself doesn't  have proper injections rather then for comments. ([link](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/yaml/injections.scm)), so I need to  to override `injections.scm` locally and add something similar to injections from markdown query
* One thing I'm afraid of - that I will need to combine parser code from `yaml` and `go-template` parsers.

**Reply by theHamsta** [reply]
Ideally, the gotmplate parser should define  ranges that are supposed to be highlighted as yaml. We would need to  have support support for `@combined`. Would it work when you inject yaml into the whole document range?

**Comment by reegnz** [comment, marked as answer]
I managed to get this far.

in ~/.config/nvim/after/plugin/treesitter.lua:
```lua
local parser_config = require'nvim-treesitter.parsers'.get_parser_configs()
parser_config.gotmpl = {
  install_info = {
    url = "https://github.com/ngalaiko/tree-sitter-go-template",
    files = {"src/parser.c"},
  },
  filetype = "gotmpl",
  used_by = {"gohtmltmpl", "gotexttmpl", "gotmpl"},
}
```

~/.config/nvim/after/ftdetect/gotmpl.vim
```vim
augroup filetypedetect
autocmd BufNewFile,BufRead *.tmpl, if search('{{.\+}}', 'nw') | setlocal filetype=gotmpl | endif
augroup END
```

I've put this into ~/.config/nvim/after/queries/gotmpl/injections.scm:
```scm
(text) @yaml
```

Also configured the highlighting queries defined at https://github.com/ngalaiko/tree-sitter-go-template#neovim-integration-using-nvim-treesitter.
Put the contents into ~/.config/nvim/after/queries/gotmpl/highlights.scm


The result looks like this (left side is TSPlayground, right side is the go template with yaml):
![Screenshot 2022-02-01 at 20 11 09](https://user-images.githubusercontent.com/5672001/152034862-da5610a5-7913-4d3f-977d-b17d9dd13ad5.png)

**Reply by reegnz** [reply]
The only thing I didn't manage to figure out yet is how to infer the language itself. Right now I needed yaml, but would be nice to infer the embedded filetype within the template.

I was thinking maybe through ftdetect (eg. `*.yaml.tmpl` for yaml, `*.html.tmpl` for html), but how do I pass that info to tree-sitter I have no idea.

**Reply by stsewd** [reply]
@reegnz you can do that with a custom directive, see this PR for an example https://github.com/nvim-treesitter/nvim-treesitter/pull/1240/files#diff-81f58ecd0f4978b575136c86619adc8dbc9a0c9e87f87702ddf22615d0ac6c4bR93

**Reply by reegnz** [reply]
Thanks for that hint @stsewd!

**Reply by joshfrench** [reply]
Did you ever get this working? I couldn't figure out how to do it.

**Reply by reegnz** [reply]
Nope, didn't spend time on doing a custom directive yet.

**Reply by ngynkvn** [reply]
Sorry to revive an old topic, but I was able to figure this out for myself, can adapt as needed for your situation.

For me, I have go templates that end like `.{ft}.tmpl`, so I added a custom directive like so:
```lua
-- init.lua --
vim.treesitter.query.add_directive("inject-go-tmpl!", function(_, _, bufnr, _, metadata)
  local fname = vim.fs.basename(vim.api.nvim_buf_get_name(bufnr))
  local _, _, ext, _ = string.find(fname, ".*%.(%a+)(%.%a+)")
  metadata["injection.language"] = ext
end, {})

vim.filetype.add({
  extension = {
    tmpl = "gotmpl",
  },
})
```

Then created a new query file to apply the custom directive:
```scm
; ~/.config/nvim/queries/gotmpl/injections.scm
((text) @injection.content
  (#inject-go-tmpl!)
  (#set! injection.combined))
```

Probably not the most reliable code, but it worked for me !

And of course, the `gotmpl` parser needs to be installed and setup to match your file extensions.

**Reply by juangiordana** [reply]
Hey @ngynkvn I just installed your plugin a couple hours ago to give it a try and so far is doing wonders. Thanks!

**Reply by Hdoc1509** [reply]
**Thanks @ngynkvn !!! It works absolutely fantastic**

 I was blowing my mind to find this solution on my own.

## BEFORE / AFTER

![inject-ejs-tmpl](https://github.com/user-attachments/assets/d44742fe-3dc3-41e0-ac79-b45ca1786e77)


### If someone is interested, [this is my nvim-config](https://github.com/Hdoc1509/nvim-config)

**Reply by ngynkvn** [reply]
I was not expecting folks to pick up on this!! So glad to hear its working for y'all :D
